### PR TITLE
Allow setting extra paths for assets:precompile and build checksum

### LINKED ISSUE #114: Doesn't recompile if including a gem that has assets
## What happened?

* What were you attempting to do?
I was adding Rails Admin to my existing application.

* What did you expect to happen?
I expected the app to recompile the assets that are now included with the addition of Rails Admin.

* What was the actual behavior? Please provide log output, if possible.
The app reused the cache, and then my app crashed when I navigated to the Rails Admin dashboard once the built container started.

## Build Configuration

* What platform (`pack`, `kpack`, `tekton` buildpacks plugin, etc.) are you
  using? Please include a version.
`pack 0.20.0+git-66a4f32.build-2668`

* What buildpacks are you using? Please include versions.
```
paketo-buildpacks/ca-certificates 2.3.2
paketo-buildpacks/mri             0.2.4
paketo-buildpacks/bundler         0.1.8
paketo-buildpacks/bundle-install  0.2.3
paketo-buildpacks/node-engine     0.6.2
paketo-buildpacks/yarn            0.4.0
paketo-buildpacks/yarn-install    0.4.0
paketo-buildpacks/rails-assets    0.2.3
paketo-buildpacks/puma            0.0.60
```

* What builder are you using? If custom, can you provide the output from `pack
  inspect-builder <builder>`?
```
full: Pulling from paketobuildpacks/builder
Digest: sha256:7a005cbd13254a3b85c75526802212d41348e4521faab26f5d2adf3565f859bf
Status: Image is up to date for paketobuildpacks/builder:full
full-cnb: Pulling from paketobuildpacks/run
Digest: sha256:fac2ca83109101e0d1dd8c26707f5f58e05ccf52b2c42db95e25bda86b9887b2
Status: Image is up to date for paketobuildpacks/run:full-cnb
```

* Can you provide a sample app or relevant configuration (`buildpack.yml`,
  `nginx.conf`, etc.)?

1. Build a basic Rails app.
2. `pack build` the app and see it succeed and run successfully.
3. Add the `rails_admin` gem to your Gemfile and install
4. Follow the directions to configure Rails Admin: https://github.com/sferik/rails_admin/wiki/Base-configuration
5. `pack build` the app and see it build successfully (assets should be reused from the cache)
6. Start the app and navigate to the admin page.
7. See the app crash with an error saying that it cannot find some required assets.

## Checklist

* [ ] I have included log output.
* [ ] The log output includes an error message.
* [x] I have included steps for reproduction.

### PR DESCRIPTION
**Pull request by bernardolins**
## Summary

Closes #114 and #241.

This change adds two new optional settings configurable through the following environment variables:

* `BP_RAILS_ASSETS_EXTRA_CHECKSUM_PATHS`:  Allows users to configure additional directories to be included in the checksum calculation during the build phase. Those directories will be included in the checksum together with the default ones. 

* `BP_RAILS_ASSETS_EXTRA_PRECOMPILE_PATHS`: Allows users to configure additional directories that will hold the results of the `assets:precompile` command. These directories' contents will be saved in the assets layer, allowing them to be utilized in future builds together with the default ones.

## Use Cases

Certain gems may utilize asset directories that differ from the default ones used in the buildpack. One example is the [vite_rails](https://rubygems.org/gems/vite_rails/versions/3.0.15) gem, which compiles the assets to the `public/vite` path. Anticipating the directories a gem might require can be difficult, so this update let users to configure the directories according to their specific needs.

## Checklist

* [x] I have viewed, signed, and submitted the Contributor License Agreement.
* [x] I have linked issue(s) that this PR should close using keywords or the Github UI (See [docs](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue))
* [x] I have added an integration test, if necessary.
* [x] I have reviewed the [styleguide](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md) for guidance on my code quality.
* [x] I'm happy with the commit history on this PR (I have rebased/squashed as needed).

### CODE REVIEW

**Comment by sophiewigmore** [comment]
@bernardolins thanks for the PR! I should have some time to review this tomorrow

**Comment by sophiewigmore** [comment]
From a preliminary read through, I support the addition of `BP_RAILS_ASSETS_EXTRA_PRECOMPILE_PATHS`, I think the use case you've laid out makes sense, I still need to test it out and review the code changes but I like it as a concept.

For `BP_RAILS_ASSETS_EXTRA_CHECKSUM_PATHS`, I also agree that this will solve the crux of the issue relating to cache reuse, but I think that the variable name is too specific to the internal buildpack implementation. Maybe something like `BP_RAILS_ASSETS_EXTRA_CACHE_PATH`? @robdimsdale do you have an opinion on the naming?

**Comment by bernardolins** [comment]
> For `BP_RAILS_ASSETS_EXTRA_CHECKSUM_PATHS`, I also agree that this will solve the crux of the issue relating to cache reuse, but I think that the variable name is too specific to the internal buildpack implementation. Maybe something like `BP_RAILS_ASSETS_EXTRA_CACHE_PATH`? 

Hi Sophie, thanks for the feedback! Your point about the checksum variable name makes sense for me, I agree that this is too specific to the implementation. I liked the variable name you suggested, let me know any other issue you find after the review!

**Comment by sophiewigmore** [comment]
Looking at issue #114, I'm actually not sure if these environment variable configurations solve the problem, but you can let me know if I'm misunderstanding since you may have more Ruby experience than me. 

The way I understood it is that when we add `rails_admin` on a second build, the assets get generated to a typical location (app/assets and app/javascript) BUT since changes don't appear until asset precompilation occurs again, no changes are available to invalidate the cache, and thus we reuse the precompile assets. As far as I can tell, these changes don't necessarily help with that issue

**Comment by robdimsdale** [comment]
I don't have a lot of rails-specific context, but on the face of it it seems surprising to have to set two different env vars. I agree with Sophie that the `BP_RAILS_ASSETS_EXTRA_CHECKSUM_PATHS` seems like the user has to know too much about the internals of the buildpack.

I can understand the use-case for adding `BP_RAILS_ASSETS_EXTRA_PRECOMPILE_PATHS`, but I'm less comfortable with having the user control caching.

Is it possible to leverage `BP_RAILS_ASSETS_EXTRA_PRECOMPILE_PATHS` for the caching, too? It's quite likely that I'm missing something, but I wanted to ask first.

**Comment by bernardolins** [comment]
Hi Rob, thanks for the review and the feedback! 

A brief recap of how the buildpack currently handles the Rails app directories.

1. The assets sources are committed to the app repo and are the base to what's going to be generated after the precompile step. The buildpack currently has a set of hardcoded paths where it looks for these asset sources and creates a checksum from them: https://github.com/paketo-buildpacks/rails-assets/blob/8d219555dac3442e0f069bb1469720818146e3a2/build.go#L83-L88

2. When `assets:precompile` is called it will generate files on different paths, and these are going to be served by the app server. The buildpack symlinks the layer to these directories: https://github.com/paketo-buildpacks/rails-assets/blob/8d219555dac3442e0f069bb1469720818146e3a2/directory_setup.go#L81-L89

The problem here is that any Gem added to the app can override the `assets:precompile` behavior, either reading from extra directories on step 1, or writing to extra directories on step 2.

The `BP_RAILS_ASSETS_EXTRA_CHECKSUM_PATHS` is meant as an extension point for the directories being read on 1. The `BP_RAILS_ASSETS_EXTRA_PRECOMPILE_PATHS` is an extension to the directories expected to contain the output from the precompile step on 2.

The only safe way I can see to drop the `BP_RAILS_ASSETS_EXTRA_CHECKSUM_PATHS` env would be by disabling the checksum step altogether. We could always restore the previous layer if we were to always run the `assets:precompile` command. The precompile command already includes an internal checksumming routine and it should take care of speeding up the build if the restored layer assets match the current source. Another option would be to add a flag that disables the checksum on the buildpack, and set this to true everytime the `BP_RAILS_ASSETS_EXTRA_PRECOMPILE_PATHS != ""`. What do you think about these options?

**Comment by robdimsdale** [comment]
Thanks for the context. That's really helpful and I appreciate you bearing with me 😅 I was hoping there would be a way to infer the checksum paths from the input but it seems like that really isn't the case.

I think a couple of integration tests (one testing each feature) would really help highlight it. Both for me, but also for future users of this feature. Is that something you feel able to tackle?

I think my main concern is that we expose the concept of `checksum` to the user. Based on the way you just described it, I view it as almost like additional input/output locations for assets. I'd love to see if we can come up with a name that abstracts over the concept of checksumming.

**Comment by bernardolins** [comment]
> I think a couple of integration tests (one testing each feature) would really help highlight it. Both for me, but also for future users of this feature. Is that something you feel able to tackle?

Sure, I think it makes sense, I can add some integration tests to this.

> I think my main concern is that we expose the concept of checksum to the user. Based on the way you just described it, I view it as almost like additional input/output locations for assets.

Yeah, I agree, based on this discussion, what do you think of changing the variable names to something like: 
`BP_RAILS_ASSETS_EXTRA_CHECKSUM_PATHS` -> `BP_RAILS_ASSETS_EXTRA_SOURCE_PATHS` 
`BP_RAILS_ASSETS_EXTRA_PRECOMPILE_PATHS` -> `BP_RAILS_ASSETS_EXTRA_DESTINATION_PATHS`

I feel that this will reflect best what those variable do without exposing internal implementation details. Let me know if you have any other suggestion, I am not too familiar with Rails nomenclatures :sweat_smile:

**Comment by sophiewigmore** [comment]
@bernardolins thanks for being so thorough in your explanations, it's been helpful.
I think `BP_RAILS_ASSETS_EXTRA_SOURCE_PATHS` and `BP_RAILS_ASSETS_EXTRA_DESTINATION_PATHS` make sense for sure, coupled with the nice description in the README I think that makes for a better user experience.

As a Rails developer, are these names meaningful to what you'd expect in to occur in your build? Obviously we might be a bit biased now since we've been doing a deep dive into these variables 😆

**Comment by sophiewigmore** [comment]
I'm still not 100% sure if this PR resolves #114 or not, since it looks like assets that get pulled in when you add `rails_admin` are in the regular location, we just don't want to reuse the cache. Am I missing anything @bernardolins?

**Comment by bernardolins** [comment]
Hi Sophie! I also appreciate the help from you and Rob, thanks for getting some time to review.

> As a Rails developer, are these names meaningful to what you'd expect in to occur in your build? Obviously we might be a bit biased now since we've been doing a deep dive into these variables 😆

I don't use Rails in a day to day basis either, but looking into the [documentation](https://guides.rubyonrails.org/asset_pipeline.html#how-to-use-the-asset-pipeline) it doesn't look like there is a specific name for those files. I can try to reach someone that has more experience with the ruby ecosystem to see if the names make sense :sweat_smile: 

> I'm still not 100% sure if this PR resolves https://github.com/paketo-buildpacks/rails-assets/issues/114 or not, since it looks like assets that get pulled in when you add rails_admin are in the regular location, we just don't want to reuse the cache. Am I missing anything @bernardolins?

Yeah, you are right, I don't know if something changed since this issue was opened, but the way it works now it doesn't look like the same problem we are trying to solve. Sorry for the mistake, I'll remove #114 from the MR description.

**Comment by sophiewigmore** [comment]
@cezarsa @bernardolins thanks for the clarification - setting `BP_RAILS_ASSETS_EXTRA_SOURCE_PATHS=Gemfile.lock` worked perfectly when I tried it. Yes it's a bit hacky, but I think I'm fine with that since we haven't heard any additional feedback from the community that this is a highly-requested feature.

**Comment by sophiewigmore** [comment]
@bernardolins the changes here look good. I have a couple small comments about test organization but overall ready to get this merged in!

File: integration/rails_6.1_test.go
Comment (sophiewigmore) [inline integration/rails_6.1_test.go:132]: Nitpick, but can we move this test into a context, and then perform this pre-test set up in an `it.Before` block? I prefer when non-test set up is separated out

File: integration/rails_7.0_test.go
Comment (sophiewigmore) [inline integration/rails_7.0_test.go:140]: Same comment as above on the 6.0 test

File: integration/reuse_layer_rebuild_test.go
Comment (sophiewigmore) [inline integration/reuse_layer_rebuild_test.go:362]: Same comment about moving test set up into an `it.Before`

Comment (robdimsdale) [reply integration/reuse_layer_rebuild_test.go:509]: Similar to my comment in the unit tests, I think it would be worth setting up a new `context` for the custom assets and moving the test setup and tests into that context.

There's no "right" answer here, as we're trying to flatten a multi-dimensional testing space (i.e. the intersection of assertions like reusing layers with features like providing custom assets) into a single, linearized graph (i.e. nested contexts). Whichever way we do this will always feel a bit clunky, and like we're repeating ourselves.

**Review by robdimsdale — COMMENTED** [review]
Generally I think this looks good. I have essentially the same requests as @sophiewigmore about moving the test code around, and a couple of (optional) refactors of the production code.

I'd be happy to merge this with just the test changes addressed.

**Comment by robdimsdale** [comment]
Looks like the tests pass but we have a single linting error. I'll try to track that down.

File: integration/reuse_layer_rebuild_test.go
Comment (robdimsdale) [inline integration/reuse_layer_rebuild_test.go:204]: This is where the linter is complaining. You can use `regexp.MustCompile` or just `Expect(err).NotTo(HaveOccurred())` as usual.

File: integration/reuse_layer_rebuild_test.go
Comment (robdimsdale) [inline integration/reuse_layer_rebuild_test.go:213]: As you refactor these tests, I would move this `err` to just directly where it's used. This works, but it's not really idiomatic with how we setup test variables. In fact, more generally, I think I'd refactor this `var` block.

These top-level `vars` construsts are generally at the top-level `context`, and they exist so that we can share references across the `Before`, `After`, and the tests themselves.

There's no need to share `err` across these closure functions, so we can just inline the declaration when we need it. Typically we have to do:

```
var err error
foo, err = <...>
```

rather than:

```
foo, err := <...>
```

to avoid shadowing `foo`. Which often is required to be a shared reference across `Before`, `After`, and the tests themselves, as mentioned above.

File: integration/rails_6.1_test.go
Comment (bernardolins) [reply integration/rails_6.1_test.go]: I moved this test case to the [integration/reuse_layer_rebuild_test.go](https://github.com/paketo-buildpacks/rails-assets/pull/348/files/ea8bc8d90eaf5ca45c427079df6ff1ba39c94d9c#diff-4972b265ee6a614ddcbcb601b874929f3b0c8aa219145dcaf51a8120c8579823) file because it makes no sense for it to be here, since the problem only happens when the app is rebuilt. 

Just for clarification: if an application uses extra destination assets paths for precompiled assets (in this test case `public/custom`) without setting the `$BP_RAILS_ASSETS_EXTRA_DESTINATION_PATHS` variable, it is possible that the build finishes and the precompiled assets are not included in the image. This is because only the default directories and those that are listed in the variable are included in the assets cache layer. 

So, if an app uses external destination paths, but doesn't set the variable:

1. in the first build, the app will run `assets:precompile`, and the assets will be available, but will not be included in the assets layer;
2.  The second build won't run the precompile command, and will link the assets saved in the assets layer to the app's working directory. Since the assets weren't cached, the build will finish successfully but the extra assets will not be included.

File: directory_setup.go
Comment (bernardolins) [reply directory_setup.go:42]: It should work, I can make a test here to see if it will cause any side effect, thanks for the suggestion

File: directory_setup_test.go
Comment (bernardolins) [reply directory_setup_test.go:47]: Yeah, it makes sense, I will improve those tests with the suggestions!

File: integration/testdata/7.0/config/application.rb
Comment (bernardolins) [reply integration/testdata/7.0/config/application.rb:26]: This one is actually not, I will remove it :)

File: integration/testdata/6.1/config/application.rb
Comment (bernardolins) [reply integration/testdata/6.1/config/application.rb:27]: Yeah, this one it is actually being used. 

This rails setting changes the path were the assets are placed after the `assets:precompile` command. By default, it will be `public/assets`, and this setting will change it to `public/custom`. It is commented because uncommenting this would break all other tests. So, during the [appropriate test case I remove this comment](https://github.com/doximity/rails-assets/blob/bernardolins-custom-assets-precompile/integration/reuse_layer_rebuild_test.go#L205). 

This is kinda hacky, but using this setting is an easy way to simulate an app that stores precompiled assets in different paths. 

I think it might be a better way to enable this without the need to keep the comment there. I will try to figure it out :)

**Comment by cezarsa** [comment]
> But, I don't know enough about Ruby to know whether this will significantly increase the rate of false-positives. That is, what proportion of changes to Gemfile.lock should result in cache invalidation? Is it high (close to 100%)? Or are there often benign changes (e.g. "last-updated-at" timestamps, or similar) that should not result in cache invalidation? I don't know. I think I'd want input from folks who actually use Ruby more than we do.

In my opinion, we should add `Gemfile.lock` to the hardcoded paths, that's the safer option since we have no way of knowing what the new Gems are doing. The lock file doesn't include timestamps or any kind of random hashes, just version numbers, and git-commit hashes if there are dependencies imported from other repositories directly. It will only change when the versions for the app dependencies really changed, how frequent that is varies greatly depending on the app though.

For the future, if we want to optimize the case where the Gem changes don't really affect assets, we could explore the option of BOTH restoring the cache AND triggering the assets:pipeline command. But that's not a code path that exists today and I think it would require more testing and validation.

File: integration/testdata/6.1/config/application.rb
Comment (robdimsdale) [reply integration/testdata/6.1/config/application.rb:27]: Ah, I hadn't made that connection.

Is it possible to append that line to the config during the test, rather than uncomment it?

If not, then I think we should add a comment to the config explaining why that line is commented out. I just worry that it will look like tech. debt and some day someone will delete it and be very confused why tests start failing.

File: directory_setup.go
Comment (robdimsdale) [reply directory_setup.go:42]: I would hope (expect?) that the existing unit tests you added should cover this - it should be a pure refactor.

**Comment by robdimsdale** [comment]
> In my opinion, we should add Gemfile.lock to the hardcoded paths, that's the safer option since we have no way of knowing what the new Gems are doing. The lock file doesn't include timestamps or any kind of random hashes, just version numbers, and git-commit hashes if there are dependencies imported from other repositories directly. It will only change when the versions for the app dependencies really changed, how frequent that is varies greatly depending on the app though.
>
> For the future, if we want to optimize the case where the Gem changes don't really affect assets, we could explore the option of BOTH restoring the cache AND triggering the assets:pipeline command. But that's not a code path that exists today and I think it would require more testing and validation.

That sounds reasonable to me. We generally hear more complaints about incorrectly using stale caches than incorrectly blowing away valid caches, so I'd be happy to lean more in this direction by default by including `Gemfile.lock` in the checksum list. I think I'd like to hear from other @paketo-buildpacks/ruby-maintainers first but I'm pretty much convinced.

File: integration/testdata/7.0/config/application.rb
Comment (sophiewigmore) [reply integration/testdata/7.0/config/application.rb:26]: It gets used in the test for the custom path, it gets uncommented

File: integration/testdata/6.1/config/application.rb
Comment (bernardolins) [reply integration/testdata/6.1/config/application.rb]: I've [changed the test](https://github.com/doximity/rails-assets/blob/bernardolins-custom-assets-precompile/integration/reuse_layer_rebuild_test.go#L200) to append a line to the `config/initializers/assets.rb` file instead, this should have the same effect without the need to keep that comment.

**Comment by bernardolins** [comment]
Hey @robdimsdale and @sophiewigmore, I think we addressed most of the comments from the last review, please let me know if there is anything else that needs improvement. Also, if you think it is a good idea, I can make a quick change to add the `Gemfile.lock` to the default checksum list.

File: directory_setup_test.go
Comment (robdimsdale) [inline directory_setup_test.go:70]: Can we use: `t.Setenv("BP_RAILS_ASSETS_EXTRA_DESTINATION_PATHS", customAssetsPrecompilePaths)` here? That way we can get rid of the `it.After` below.

File: directory_setup_test.go
Comment (robdimsdale) [inline directory_setup_test.go:108]: Same as above - can we use `t.Setenv` here?

File: directory_setup_test.go
Comment (robdimsdale) [inline directory_setup_test.go:156]: Same as above - can we use `t.Setenv` here?

**Comment by sophiewigmore** [comment]
@robdimsdale @bernardolins @cezarsa I'm not entirely sure how I feel about adding the Gemfile.lock into the checksum list, because I don't know if there are further ramifications to blowing the cache away on Gemfile.lock changes we might not be considering. I think I need to do a bit more research to fully get on board, I'd be open to adding it eventually, maybe in a separate PR when I can do some investigation?
I don't want to block this PR any further by waiting for full buy-in on that change, if that makes sense

**Review by robdimsdale — COMMENTED** [review]
Thanks for turning this around. Just a couple of minor test improvements, and then I think we'll be good to go!

**Comment by robdimsdale** [comment]
I would suggest creating a separate PR to add `Gemfile.lock` to the default list.

That way we can more easily discover that change and why we did it, and more easily revert it if necessary.

**Comment by sophiewigmore** [comment]
@robdimsdale not sure if you saw my latest comment, as it looks like we posted at the same time. :)

**Comment by bernardolins** [comment]
> I'm not entirely sure how I feel about adding the Gemfile.lock into the checksum list, because I don't know if there are further ramifications to blowing the cache away on Gemfile.lock changes we might not be considering. I think I need to do a bit more research to fully get on board, I'd be open to adding it eventually, maybe in a separate PR when I can do some investigation?
I don't want to block this PR any further by waiting for full buy-in on that change, if that makes sense

That sounds good for me, once we are done with this MR I can open another one adding the Gemfile.lock so we can test it properly. We are planning to include this file to the `$BP_RAILS_ASSETS_EXTRA_SOURCE_PATHS` in our production setup, so I can report back any problem we find.

**Review by robdimsdale — APPROVED** [review]
This looks great!

Thanks for keeping on top of this PR and accommodating our requests. I think this code looks really good and I'm happy to merge it.

**Comment by robdimsdale** [comment]
@bernardolins I think you'll need to rebase this (or give the `paketo-buildpacks/ruby-maintainers` access to your fork) before we can merge it.

**Comment by bernardolins** [comment]
> @bernardolins I think you'll need to rebase this (or give the `paketo-buildpacks/ruby-maintainers` access to your fork) before we can merge it.

Hey Rob, I just rebased the branch, let me know if there is anything else I can do.

> This looks great!
> 
> Thanks for keeping on top of this PR and accommodating our requests. I think this code looks really good and I'm happy to merge it.

Thank you for the feedback and also thanks for the help from you and Sophie during the whole process :smile:

**Comment by robdimsdale** [comment]
@bernardolins no problem, but the branch is still out of date. I've enabled auto-merge, so once you rebase it, it should merge after the tests pass.

**Pull request merged**
